### DISCUSSION TITLE
Selecting MCU for a new keyboard

### DISCUSSION THREAD

**Discussion by Looooopy** [joshreve/dactyl-keyboard#60] [Q&A] [answered]
Just looking on options on MCU.

I have seen a lot of Arduino Pro Micros and some Teensy 2.0 processors utilized in keyboards but is that enough, I suppose it depends on what you are thinking of doing with your keyboard. But have anyone of you who using those MCU noticed any downside or is it working flawless?

So if you guys would build a new keyboard today which MCU would you pick
* Teensy 2.0 (Supported by QMK) - Hard for me to get in Europe
* Pro Micro (Supported by QMK) - Have a couple laying around
* ESP32 (NOT supported by QMK [Open Issue](https://github.com/qmk/qmk_firmware/issues/9223)) - Have a couple of these laying around, but it's not an option to start with it because it would require a lot of fiddling.
* STM32 (Supported by QMK) - Need to get some because all are in my drones :)

Did I miss any MCU that are good to use for keyboards?

**Comment by joshreve** [comment, marked as answer]
I want to move to a pi pico, adafruit feather, or other ARM chips for the speed, extended I/O, and added memory so a firmware library can focus on readability and interface rather than maximum efficiency.  

I am also interested in seeing the Arduino libraries used for hardware abstraction so new MCUs are easier to change.  I think we are far off of this, but I can hope.  My biggest issue with QMK is that it attempts low level interfaces with everything rather than using clearly defined abstraction layers.  QMK is impressive and has the most features of the firmware options, but it is challenging to add new hardware or features because of the SW architecture.

I feel like Arduino already handles this and will for new hardware for the foreseeable future and it is a shame that is not being utilized on keyboards.

For now though, I find promicros sufficient, but they limit you on number of luxury features like OLED, trackball, etc.

**Reply by Looooopy** [reply]
I stumbled low memory on the Pro micro with OLED exactly like you described so I ordered a couple of STM32F411 which would probably solve my issues.

I have to turn off things that I am not testing for now so I can focus on the pages on my OLED and it feels like am soon turning off features that I need for the actual test :)

**Reply by Looooopy** [reply]
>I would now Recommend all people that use any other feature that is not just keyboard to get a STM32 processor with more memory. if not just to future proof it.

**Reply by Looooopy** [reply]
Found an alternative solution for Arduino that get you a bit on the way I got 3,3KiB free after turning on this feature and every feature you turn on later also takes up less space.

`LTO_ENABLE = yes`

Read more here
https://docs.qmk.fm/?id=squeezing-the-most-out-of-avr#/squeezing_avr?id=rulesmk-settings

**Reply by xisang233** [reply]
> I want to move to a pi pico, adafruit feather, or other ARM chips for the speed, extended I/O, and added memory so a firmware library can focus on readability and interface rather than maximum efficiency.
> 
> I am also interested in seeing the Arduino libraries used for hardware abstraction so new MCUs are easier to change. I think we are far off of this, but I can hope. My biggest issue with QMK is that it attempts low level interfaces with everything rather than using clearly defined abstraction layers. QMK is impressive and has the most features of the firmware options, but it is challenging to add new hardware or features because of the SW architecture.
> 
> I feel like Arduino already handles this and will for new hardware for the foreseeable future and it is a shame that is not being utilized on keyboards.
> 
> For now though, I find promicros sufficient, but they limit you on number of luxury features like OLED, trackball, etc.

I think rp2040 is a good choice. It has megabytes ROM and impressive PIO.
But we have to make a controller holder for it because it's much more bigger than pro micro. I remade one on my 5x7 keyboard and it works but I wonder if it's too big for any smaller keyboard.
